example/grammar/magnet.cpp

```diff
@@ -0,0 +1,496 @@
+//
+// Distributed under the Boost Software License, Version 1.0.
+// https://www.boost.org/LICENSE_1_0.txt
+//
+
+#include <boost/url/url_view.hpp>
+#include <boost/url/rfc/absolute_uri_rule.hpp>
+#include <boost/url/grammar/charset.hpp>
+#include <iostream>
+
+using namespace boost::urls;
+
+// A small filter and transform view
+template <class C, class V, class P, class T>
+class filter_view
+{
+    C c_;
+    P p_;
+    T t_;
+
+    using value_type = V;
+
+public:
+    class iterator
+    {
+        friend filter_view;
+
+        using base_iter = typename C::iterator;
+
+        base_iter b_;
+        filter_view& v_;
+
+        void
+        increment_false()
+        {
+            auto e = v_.c_.end();
+            while (b_ != e && !v_.p_(*b_))
+                ++b_;
+        }
+
+        iterator(
+            filter_view& v,
+            base_iter it)
+            : b_(it)
+            , v_(v)
+        {
+            increment_false();
+        }
+
+    public:
+        iterator&
+        operator++() noexcept
+        {
+            ++b_;
+            increment_false();
+            return *this;
+        }
+
+        iterator
+        operator++(int) noexcept
+        {
+            auto tmp = *this;
+            ++*this;
+            return tmp;
+        }
+
+        value_type
+        operator*() const
+        {
+            return v_.t_(*b_);
+        }
+
+        friend
+        bool
+        operator==(
+            iterator const& a,
+            iterator const& b) noexcept
+        {
+            return a.b_ == b.b_;
+        }
+
+        friend
+        bool
+        operator!=(
+            iterator const& a,
+            iterator const& b) noexcept
+        {
+            return a.b_ != b.b_;
+        }
+    };
+
+    filter_view(
+        C const& base,
+        P pred = {},
+        T transform = {})
+        : c_(base)
+        , p_(pred)
+        , t_(transform) {};
+
+    iterator begin()
+    {
+        return {*this, c_.begin()};
+    }
+
+    iterator end()
+    {
+        return {*this, c_.end()};
+    }
+};
+
+struct magnet_link_rule_t;
+
+// A new url type for magnet links
+class magnet_link_view
+{
+    url_view u_;
+
+    struct is_exact_topic {
+        bool operator()(query_param_encoded_view p)
+        {
+            if (p.key == "xt")
+                return true;
+            return
+                p.key.starts_with("xt.") &&
+                p.key.size() > 3 &&
+                p.key[2] == '.' &&
+                std::all_of(
+                    p.key.begin() + 3,
+                    p.key.end(),
+                    grammar::digit_chars);
+        }
+    };
+
+    class is_url_with_key {
+        string_view k_;
+
+    public:
+        explicit
+        is_url_with_key(string_view key) : k_(key) {}
+
+        bool operator()(query_param_encoded_view p)
+        {
+            if (p.key != k_)
+                return false;
+            error_code ec;
+            char buf_[1024];
```

Comment:
eww... a hardcoded limit? Not feeling it..

Reply:
Yes. This is the least bad solution I found for something peculiar about the syntax. We have to check if this is a valid URL. The problem is the URL is percent-encoded twice. 

So `parse_uri(string_view)` doesn't work and `parse_uri(pct_encoded_view)` doesn't exist. `parse_uri(pct_encoded_view)` doesn't even make sense because it can't even refer to anything. 

The only complete solution would be another rule for a "url_view" that would point to a string that has been percent-encoded twice, but that's not worth the example.

Reply:
Another solution would be just to not check if the value is a valid url. Then the "user" is responsible for doing that.

Reply:
This points to a flaw in our design. Not sure if it is worth fixing now (or later).

Reply:
We could require that the user pass in a _MutableString_ as a temporary buffer used to first hold the percent-decoded string, and then in the call to `parse_uri`. The user would receive a `url_view` which points to the mutable string buffer. The user can then decide if they want to keep the url view, or if they want to construct an owning `url` from it.